src/java/io/bazel/rulesscala/worker/Worker.java

```diff
@@ -78,37 +74,64 @@ public void checkPermission(Permission permission) {
             break;
           }
 
-          int code = 0;
-
-          try {
-            workerInterface.work(stringListToArray(request.getArgumentsList()));
-          } catch (ExitTrapped e) {
-            code = e.code;
-          } catch (Exception e) {
-            System.err.println(e.getMessage());
-            e.printStackTrace();
-            code = 1;
+          if (request.getRequestId() == 0) {
+            processWorkRequest(workerInterface, request);
+          } else {
+            executorService.submit(
+                () -> {
+                  processWorkRequest(workerInterface, request);
+                });
           }
 
-          WorkerProtocol.WorkResponse.newBuilder()
-              .setExitCode(code)
-              .setOutput(outStream.toString())
-              .build()
-              .writeDelimitedTo(stdout);
-
         } catch (IOException e) {
           // for now we swallow IOExceptions when
-          // reading/writing proto
-        } finally {
-          out.flush();
-          outStream.reset();
-          System.gc();
+          // reading proto
         }
       }
     } finally {
       System.setIn(stdin);
-      System.setOut(stdout);
-      System.setErr(stderr);
+    }
+  }
+
+  private static Object lock = new Object();
+
+  private static void processWorkRequest(
+      Interface workerInterface, WorkerProtocol.WorkRequest request) {
+    int code = 0;
+    ByteArrayOutputStream outStream = new ByteArrayOutputStream();
+    PrintStream out = new PrintStream(outStream);
+
+    try {
+      workerInterface.work(stringListToArray(request.getArgumentsList()), out, out);
+    } catch (ExitTrapped e) {
+      code = e.code;
+    } catch (Exception e) {
+      out.println(e.getMessage());
+      e.printStackTrace(out);
+      code = 1;
+    }
+
+    try {
+      out.flush();
+      WorkerProtocol.WorkResponse response =
+          WorkerProtocol.WorkResponse.newBuilder()
+              .setExitCode(code)
+              .setOutput(outStream.toString())
+              .setRequestId(request.getRequestId())
+              .build();
+
+      synchronized (lock) {
```

Comment:
Maybe it would be possible to synchronize on `System.out`? That would reveal intent better. Or maybe syncronozed method? Anyway lock object is also fine.

Reply:
I read that having a private lock object is the safer way of doing synchronization, because locking on a public object like System.out might create dead locks if someone else does the same.